test/suite/test_hs22.py

```diff
@@ -0,0 +1,121 @@
+# ARISING FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR
+# OTHER DEALINGS IN THE SOFTWARE.
+
+import wiredtiger, wttest
+
+def timestamp_str(t):
+    return '%x' % t
+
+# test_hs22.py
+# Test we don't crash when the out of order timestamp update is followed by a tombstone.
+class test_hs22(wttest.WiredTigerTestCase):
+    conn_config = 'cache_size=50MB'
+    session_config = 'isolation=snapshot'
+
+    def test_onpage_out_of_order_timestamp_update(self):
+        uri = 'table:test_hs22'
+        self.session.create(uri, 'key_format=S,value_format=S')
+        cursor = self.session.open_cursor(uri)
+        self.conn.set_timestamp(
+            'oldest_timestamp=' + timestamp_str(1) + ',stable_timestamp=' + timestamp_str(1))
+
+        value1 = 'a'
+        value2 = 'b'
+
+        # Insert a value
+        self.session.begin_transaction()
+        cursor[str(0)] = value1
+        self.session.commit_transaction('commit_timestamp=' + timestamp_str(10))
+
+        # Remove the value
+        self.session.begin_transaction()
+        cursor.set_key(str(0))
+        self.assertEqual(cursor.remove(), 0)
+        self.session.commit_transaction('commit_timestamp=' + timestamp_str(20))
+
+        # Do an out of order timestamp update
+        self.session.begin_transaction()
+        cursor[str(0)] = value2
+        self.session.commit_transaction('commit_timestamp=' + timestamp_str(15))
+
+        # Insert another key
+        self.session.begin_transaction()
+        cursor[str(1)] = value1
+        self.session.commit_transaction('commit_timestamp=' + timestamp_str(20))
+
+        # Update the key
+        self.session.begin_transaction()
+        cursor[str(1)] = value2
+        self.session.commit_transaction('commit_timestamp=' + timestamp_str(30))
+
+        # Do a checkpoint to trigger history store reconciliation
+        self.session.checkpoint()
+
+    def test_out_of_order_timestamp_update_newer_than_tombstone(self):
+        uri = 'table:test_hs22'
+        self.session.create(uri, 'key_format=S,value_format=S')
+        cursor = self.session.open_cursor(uri)
+        self.conn.set_timestamp(
+            'oldest_timestamp=' + timestamp_str(1) + ',stable_timestamp=' + timestamp_str(1))
+
+        value1 = 'a'
+        value2 = 'b'
+
+        # Insert a value
+        self.session.begin_transaction()
+        cursor[str(0)] = value1
+        self.session.commit_transaction('commit_timestamp=' + timestamp_str(10))
+
+        # Remove the value
+        self.session.begin_transaction()
+        cursor.set_key(str(0))
+        self.assertEqual(cursor.remove(), 0)
+        self.session.commit_transaction('commit_timestamp=' + timestamp_str(20))
+
+        # Do an out of order timestamp update
+        self.session.begin_transaction()
+        cursor[str(0)] = value2
+        self.session.commit_transaction('commit_timestamp=' + timestamp_str(15))
+
+        # Add another update
```

Comment:
I don't clearly understand the difference between this test and the first one expect an extra insert. Can you clarify the test?

Reply:
The second one the out of order timestamp is not the onpage value.